apps/accounts/views.py

```diff
@@ -72,72 +99,100 @@ def login_user(request, user, identity):
-def with_user(fn):
-    @functools.wraps(fn)
-    def inner(self, request):
-        data = request.GET if request.method == 'GET' else request.DATA
-        state_parts = data.get('state', '').split(':', 1)
-        state = state_parts[0]
-        if 'code' not in data:
-            log.info('No code provided.')
-            return Response({'error': 'No code provided.'}, status=422)
-        elif (not request.session.get('fxa_state') or
-                request.session['fxa_state'] != state):
-            log.info('State mismatch. URL: {url} Session: {session}'.format(
-                url=data.get('state'),
-                session=request.session.get('fxa_state'),
-            ))
-            return Response({'error': 'State mismatch.'}, status=400)
+def render_error(request, error, next_path=None, format=None):
+    if format == 'json':
+        status = ERROR_STATUSES[error]
+        return Response({'error': error}, status=status)
+    else:
+        if not is_safe_url(next_path):
+            next_path = None
+        return HttpResponseRedirect(
+            urlparams(reverse('users.login'), error=error, to=next_path))
+
 
+def parse_next_path(state_parts):
+    next_path = None
+    if len(state_parts) == 2:
+        # The = signs will be stripped off so we need to add them back
+        # but it only cares if there are too few so add 4 of them.
+        encoded_path = state_parts[1] + '===='
         try:
-            identity = verify.fxa_identify(data['code'],
-                                           config=settings.FXA_CONFIG)
-        except verify.IdentificationError:
-            log.info('Profile not found. Code: {}'.format(data['code']))
-            return Response({'error': 'Profile not found.'}, status=401)
-        else:
-            identity_user = find_user(identity)
-            if request.user.is_authenticated():
-                if identity_user is not None and identity_user != request.user:
-                    log.info('Conflict finding user during FxA login. '
-                             'request.user: {}, identity_user: {}'.format(
-                                 request.user.pk, identity_user.pk))
-                    return Response({'error': 'User mismatch.'}, status=422)
-                elif request.user.fxa_migrated():
-                    log.info('User already migrated. '
-                             'request.user: {}, identity_user: {}'.format(
-                                 request.user, identity_user))
-                    return Response(
-                        {'error': 'User already migrated.'}, status=422)
-                else:
-                    user = request.user
+            next_path = base64.urlsafe_b64decode(str(encoded_path))
+        except TypeError:
+            log.info('Error decoding next_path {}'.format(
+                encoded_path))
+            pass
+    if not is_safe_url(next_path):
+        next_path = None
+    return next_path
+
+
+def with_user(format=None):
```

Comment:
curl sets `*/*` by default. Rest framework 2 doesn't seem to have a good way of detecting the expected format and parsing accept headers accurately seems like a real pain.

Rest framework 2 supports renderers but since this isn't actually rendering HTML it's redirecting off of this URL this can't use a renderer.

This seemed like the most straightforward solution. I removed the default value for `format` here so it will blow up at load time if it isn't set.